loader/src/reference_binder.py

```diff
@@ -0,0 +1,150 @@
+                {"identifier": {"$elemMatch": identifier}}, ["id"],
+            )
+            if referenced_resource:
+                # if found, add the ID as the "literal reference"
+                # (https://www.hl7.org/fhir/references-definitions.html#Reference.reference)
+                logger.info(f"reference to {reference_type} {identifier} resolved")
+                ref["reference"] = f"{reference_type}/{referenced_resource['id']}"
+            else:
+                logger.info(
+                    f"caching reference to {reference_type} {identifier} at {reference_path}"
+                )
+
+                # otherwise, cache the reference to resolve it later
+                target_ref = (reference_type, identifier_tuple)
+                source_ref = (fhir_object["resourceType"], reference_path, isArray)
+                self.cache[target_ref][source_ref].append(fhir_object["id"])
+            return ref
+
+        # If we have a list of references, we want to bind all of them.
+        # Thus, we loop on all the items in reference_attribute.
+        if isinstance(reference_attribute, list):
+            return [bind(ref, isArray=True) for ref in reference_attribute]
+        else:
+            return bind(reference_attribute)
+
+    def resolve_pending_references(self, fhir_object):
+        for identifier in fhir_object["identifier"]:
+            try:
+                identifier_tuple = self.extract_key_tuple(identifier)
+            except Exception as e:
+                logger.error(e)
+                continue
+
+            target_ref = (fhir_object["resourceType"], identifier_tuple)
+            pending_refs = self.cache.get(target_ref, {})
+            for (source_type, reference_path, isArray), refs in pending_refs.items():
+                find_predicate = {
+                    "id": {"$in": refs},
+                    #
+                    reference_path: {"$elemMatch": {"identifier": identifier}}
+                    if isArray
+                    else {"identifier": identifier},
+                }
+                # handle updating reference arrays:
+                # we keep the indices in the path (eg: "identifier.0.assigner.reference")
+                # but if fhir_object[reference_path] is an array, we use the '$' feature of mongo
+                # in order to update the write element of the array.
+                # https://docs.mongodb.com/manual/reference/operator/update/positional/#update-documents-in-an-array
+                update_predicate = {
+                    "$set": {
+                        f"{reference_path}{'.$' if isArray else ''}.reference": fhir_object["id"]
+                    }
+                }
+                logger.info(
+                    "Updating resource %s: %s %s", source_type, find_predicate, update_predicate
+                )
+                self.fhirstore.db[source_type].update_many(find_predicate, update_predicate)
```

Comment:
It shouldn't happen if users don't do random stuff (or maybe it could in some weird cases?) but will this work if several elements of an array of References are pointing to the current document?

Reply:
no, if multiple elements match the query `{reference_path: {"$elemMatch": partial_identifier(identifier) } }` then only the first one will be updated. I added a comment on that